compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1030,24 +1061,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let count = self.to_const(count);
 
         let uty = match expected {
-            ExpectHasType(uty) => match *uty.kind() {
+            ExpectHasType(uty, _) => match *uty.kind() {
                 ty::Array(ty, _) | ty::Slice(ty) => Some(ty),
                 _ => None,
             },
             _ => None,
         };
 
+        let type_ascr_ctxt = expected.get_coercion_ctxt();
         let (element_ty, t) = match uty {
             Some(uty) => {
-                self.check_expr_coercable_to_type(&element, uty, None);
+                self.check_expr_coercable_to_type(&element, uty, None, type_ascr_ctxt);
```

Comment:
as above, this argument should not be necessary, but this should always be a coercion site